### ISSUE TITLE
Automatically search for grammar when not found

### ISSUE THREAD

**Issue by mickaelistria** [eclipse/tm4e#157]
GitHub textmate organization is actually a repo of TextMate grammars. When a file is opened with Generic Editor and expects TM4E to participate but tm4e has no grammar, it should open a dialog to user suggesting to search for grammars on GitHub and allow to fetch them. A query like `"<string>cpp</string>" user:textmate repo:*.tmbundle extension:plist` (in case of .cpp) can for example show interesting results.

**Comment by mickaelistria** [comment]
GitHub does allow to access raw content of repos through http: https://raw.githubusercontent.com/textmate/c.tmbundle/master/Syntaxes/C.plist

**Comment by angelozerr** [comment]
Yes you are right, but how we can map a textmate grammar with this URL? I mean genericeditor opens a file with extends `.c`, how we can discover the well URL https://raw.githubusercontent.com/textmate/c.tmbundle/master/Syntaxes/C.plist? Perhaps we should have a textmate serach engine  which does that?

**Comment by angelozerr** [comment]
An idea is to have a server like https://tmsearch.com/ which does the same thing for npm https://npmsearch.com/ (to find package)

**Comment by angelozerr** [comment]
@ainslec perhaps this textmate search engine could b ean issue for https://eeyo.io/iro/ ? What do you think about that?

**Comment by mickaelistria** [comment]
> but how we can map a textmate grammar with this URL?

We don't map it, we send the search request to GitHub and tell user "We found this grammar from GitHub, do you want to use it?". No additional registry involved, GitHub is the registry and search engine.

**Comment by angelozerr** [comment]
> "We found this grammar from GitHub, do you want to use it?"

Sorry I have not understood your idea?

An idea is that we could use teh APi Search of github https://developer.github.com/v3/search/

For instance:

 *  we can retrieve list of github repositories with https://api.github.com/search/repositories?q=tmBundle

```json
{
  "total_count": 1673,
  "incomplete_results": false,
  "items": [
    {
      "id": 793799,
      "name": "SCSS.tmbundle",
      "full_name": "MarioRicalde/SCSS.tmbundle",
....

```

 * after we can retrieve the URL like this https://api.github.com/search/code?q=repo:MarioRicalde/SCSS.tmbundle+extension:tmLanguage

**Comment by ainslec** [comment]
@angelozerr 

> perhaps this textmate search engine could b ean issue for https://eeyo.io/iro/ ? What do you think about that?

What kind of feature do you want it to be? I understand the concept of a .tmLanguage registry, but what role do you think [Iro](https://eeyo.io/iro) should play? Should it add an option to publish to a central registry, or provide a registry itself? Just trying to understand your comment.

**Comment by angelozerr** [comment]
> What kind of feature do you want it to be?

Just having a textmate search engine like search textmate grammar .tmLanguage by file extension.

**Comment by ainslec** [comment]
Problem is that Iro is a static web page, nothing more. And it doesn't store anything on the serverside (because there is no serverside).

**Comment by angelozerr** [comment]
> Seems like the problem is not a tech problem, but a hosting problem. Maybe the Eclipse project can host the search engine?

I think we can consume the Github API Search for that even if it's not perfect, but IMHO I think the better clean solution is to have a search engine like we have for npm.

**Comment by angelozerr** [comment]
> A query like "<string>cpp</string>" user:textmate repo:*.tmbundle extension:plist (in case of .cpp) can for example show interesting results.

Could you give me the right github URL which retrieves this grammar please?

**Comment by mickaelistria** [comment]
I'm going submarine and doing only bugfix, releng or consolidation of existing features for Photon. I'll come back to this topic in a few weeks as discovery (of Docker based LS, or TM grammars) is likely to become a hot topic for me.

**Comment by angelozerr** [comment]
Ok I wanted to start something, but if you have the intetion to do it, I let you implement this feature.
